### DISCUSSION TITLE
Customer Data error for ParadoxLabs Authnetcim plugin

### DISCUSSION THREAD

**Discussion by delfina-rossi** [breezefront/community#330] [1. Help]
After fixing an issue with region-updater.js for this plugin I'm running into another error.

This plugin for authorize.net allows to save a credit card in "My Payment Methodths" under "My Account" . After completing credit card information and clicking on submit, saving the credit card fails.
It works under Luma so it's a breeze integration issue.
It seems to break at function from line 284  in customer-data.js

 ```
$(document).on('submit', function (event) {
        var names;

        if (!event.target.method.match(/post|put|delete/i)) {
            return;
        }

        names = sectionConfig.getAffectedSections(event.target.action);

        if (!names.length) {
            return;
        }

        customerData.invalidate(names);
        $.storage.remove(names);
    });
```
giving this error:

![error-customer-data](https://github.com/user-attachments/assets/75af76ce-dd80-4472-b784-b0cd0b976372)

Any help or clue to get this solved?

**Comment by vovayatsyuk** [comment]
I wonder why it's working on Luma without errors :thinking:

Looks like this line was recently updated in Luma, but not released yet: https://github.com/magento/magento2/commit/f3c3f7ee866cff9c13c8e91e81e265a85a4433d6

I've just applied the same fix in Breeze: https://github.com/breezefront/module-breeze/commit/1bb89757540f1e0213a74920a28adb87202605e9

**Comment by delfina-rossi** [comment]
thank you Vova. That error seems to be gone now, I'm struggling though to make the rest work.
My goal is to make this Paradox Labs authorize.net plugin compatible and it's getting very difficult.

I'm reading the documentation, I have a quick example now see if I understand correctly.

In order to make this cardFormatter component work, I've added this line:


     `   component: 'ParadoxLabs_TokenBase/js/cardFormatter', // This line is required in Breeze`

here:

```
define([
    "jquery",
    'Magento_Payment/js/model/credit-card-validation/credit-card-number-validator'
], function($, cardNumberValidator) {
    "use strict";


    $.widget('mage.tokenbaseCardFormatter', {
        component: 'ParadoxLabs_TokenBase/js/cardFormatter', // This line is required in Breeze
        options: {
            ccInputSelector: '[autocomplete="cc-number"]',
            separator: ' ',
            ccTypeSelector: '',
            ccTypeContainer: ''
        },

        _create: function() {
            var ccInput = this.element.find(this.options.ccInputSelector);

            if (ccInput.length > 0) {
                ccInput.on('keydown', this.handleKeydown.bind(this))
                       .on('input paste', this.formatCc.bind(this));
                console.log('si soy');
            }

            if (this.options.ccTypeSelector !== '' && this.options.ccTypeContainer !== '') {
                ccInput.on('input change keyup paste', this.detectCcType.bind(this));
            }
        },

        /**
         * Track cursor through the separator for deletes, etc.
         */
        handleKeydown: function(e) {
            var ccInput = this.element.find(this.options.ccInputSelector)[0];
            var cursor = ccInput.selectionStart;

            if (ccInput.selectionEnd !== cursor) {
                return;
            }

            if (e.which === 46 && ccInput.value[cursor] === this.options.separator) {
                ccInput.selectionStart++;
            } else if (e.which === 8 && cursor && ccInput.value[cursor - 1] === this.options.separator) {
                ccInput.selectionEnd--;
            }
        },

        /**
         * Format the CC number element with nice separators.
         */
        formatCc: function() {
            var ccInput = this.element.find(this.options.ccInputSelector)[0];
            var value   = ccInput.value;
            var cursor  = ccInput.selectionStart;

            if (value.substring(0,4) === "XXXX") {
                return;
            }

            var matches = value.substring(0, cursor).match(/[^0-9]/g);
            if (matches) {
                cursor -= matches.length;
            }

            value = value.replace(/[^0-9]/g, "").substring(0, 19);

            // AmEx cards get separated different, because they're special.
            var typeIsAmEx = false;
            if (value.substring(0,2) === "34" || value.substring(0,2) === "37") {
                typeIsAmEx = true;
            }

            var formatted = "";
            for (var i=0, n=value.length; i<n; i++) {
                if (this.shouldSeparate(i, typeIsAmEx)) {
                    if (formatted.length <= cursor) {
                        cursor++;
                    }

                    formatted += this.options.separator;
                }

                formatted += value[i];
            }

            if (formatted === ccInput.value) {
                return;
            }

            ccInput.value = formatted;
            ccInput.selectionEnd = cursor;
        },

        /**
         * Determine whether a card number should be separated at the given index.
         *
         * @param int i
         * @param bool isAmEx
         * @returns bool
         */
        shouldSeparate: function(i, isAmEx) {
            // Separate AmEx cards at 4 and 10 rather than quadruplets.
            if (isAmEx === true) {
                return i === 4 || i === 10;
            }

            return i && i % 4 === 0;
        },

        /**
         * Detect CC type as number is entered, and update active type.
         */
        detectCcType: function() {
            this.updateCcType(null);

            var cc_no = this.element.find(this.options.ccInputSelector).val().replace(/\D/g,'');
            var result = cardNumberValidator(cc_no);

            if ((result.isPotentiallyValid || result.isValid) && result.card !== null) {
                this.updateCcType(result.card.type);
            }
        },

        /**
         * Update active CC type to the selected value.
         *
         * @param string creditCardType
         */
        updateCcType: function (creditCardType) {
            this.element.find(this.options.ccTypeSelector).val(creditCardType).trigger('change');

            var typeContainer = $(this.element).find(this.options.ccTypeContainer);
            typeContainer.find('._active').removeClass('_active');
            typeContainer.find('[data-type=' + creditCardType + ']').addClass('_active');
        }
    });

    return $.mage.tokenbaseCardFormatter;
});

```

am I supposed to do something else than adding that line? I'm lost on how to integrate it properly

**Reply by vovayatsyuk** [reply]
1. You need to create `breeze_default.xml` with component registration. After this step, if done properly, the file should be downloaded by Breeze if this component found on the page. Use developer toolbar to check if this file gets downloaded.
2. Then you need to add `component: ...`. After this step, if done properly, breeze will execute `_create` method if this component found on the page.

Can you share the link to your integration repo? It would be much easier to help in this case.

**Comment by delfina-rossi** [comment]
Hi Vova, first of all thank you for your help. I'm very sorry I cannot share code in an easier way for you to help, I still appreciate it.
Here are some screenshots, I did create breeze_default.xml with no luck. The js is called under ParadoxLabs/Authnetcim but the js is placed in ParadoxLabs/TokenBase. Could the location affect? I've registered breeze_default in ParadoxLabs/TokenBase only

ParadoxLabs/TokenBase/view/base/layout/breeze-default.xml
![breeze-default](https://github.com/user-attachments/assets/67fdea4d-372c-488c-9c9d-8ca5017ae9d9)

ParadoxLabs/TokenBase/view/base/web/js/cardFormatter.js
![breeze-component](https://github.com/user-attachments/assets/3e034921-d786-465f-828a-81406f4d295f)

ParadoxLabs/Authnetcim/view/frontend/templates/customer/form/cc.phtml
![breeze-use](https://github.com/user-attachments/assets/39ebc4fb-9657-4294-afbf-063ec6700c10)

**Reply by vovayatsyuk** [reply]
If I understand properly ParadoxLabs uses `tokenbaseCardFormatter` name for its component in `cc.phtml`

Try changing the **name** in breeze_default.xml from `ParadoxLabs_TokenBase/js/cardFormatter` to `tokenbaseCardFormatter`.

**Reply by delfina-rossi** [reply]
tried with no luck, first only changing name in breeze_default.  Then I also changed the component name in cardFormatter.js.
None worked :(

**Reply by vovayatsyuk** [reply]
I just tried to do the same, and it works fine:

My phtml:

```html
<script type="text/x-magento-init">
    {
        "[data-gallery-id='<?= $galleryId ?>']": {
            "mage/gallery/gallery": {
                "data": <?= /* @noEscape */ $block->getGalleryImagesJson() ?>,
                "options": <?= /* @noEscape */ $block->getGalleryOptions()->getOptionsJson() ?>,
                "magnifierOpts": <?= /* @noEscape */ $block->getMagnifier() ?>
            },
            "tokenbaseCardFormatter": {}
        }
    }
</script>
```

My XML:

```xml
<item name="default" xsi:type="array">
    <item name="items" xsi:type="array">
        <item name="tokenbaseCardFormatter" xsi:type="array">
            <item name="path" xsi:type="string">Swissup_Breeze/js/components/aaa</item>
        </item>
...
```

aaa.js - is the copy of your js script.

![image](https://github.com/user-attachments/assets/21ab27b2-86a1-4f5b-b7b5-2c97d2a0645f)

**Comment by delfina-rossi** [comment]
Hi @vovayatsyuk I still have no luck making this work, there's probably something else missing.
This is the full plugin https://store.paradoxlabs.com/magento2-authorize-net-cim-payment-module.html
can you help get it integrated properly?
It works for payments @ checkout. The only thing not working  is under my account / my payment options the ability to add a new card there. It doesn't work, but like I mentioned it works fine at checkout.
And it works properly in luma theme.
Thank's in advance in case you can give it a try.

**Reply by vovayatsyuk** [reply]
I hope this module will help you with other integrations: https://github.com/breezefront/module-breeze-paradoxlabs-authnetcim

I had to update the breeze module a bit, to fix an issue with the number validation in credit card field: https://github.com/breezefront/module-breeze/commit/0e2c8c5887d783848ac80b4bbc55d4be5595b5f8

**Reply by delfina-rossi** [reply]
Hi Vova, thank you for your assitance.
It's definetely better. cardFormatter.js is now working, but the functionality of saving the card is still not working. Where you able to test the Paradox plugin? From what I can see, accept.js is also needed, but adding that to breeze_default is still not enough to make the whole thing work.

Adding a card on this screen is the one failing (works on luma) in case you can try the integration:

![payment-options](https://github.com/user-attachments/assets/68f338e1-e89c-4eb0-ba0c-b032a2e9ba6c)

**Reply by vovayatsyuk** [reply]
Can you record the screen with opened dev console? For me everything was working fine and I was able to add 2 cards.

**Reply by delfina-rossi** [reply]
sorry, my bad. I did composer update for swissup/module-breeze hoping to get that last commit you mentioned and I've just realized I did not receive that, so first I'll make sure my breeze code is updated properly.

Should composer update be enough to get last code?

**Reply by vovayatsyuk** [reply]
No, since it wasn't released yet. I recommend applying it manually.

**Reply by delfina-rossi** [reply]
Ok, I applied it manually and tested but cannot get it to work

https://watch.screencastify.com/v/siZbYnCqp1vMbxgtypvd

tested it on original breeze blank and I also get the "Accept is not defined" error.

**Reply by vovayatsyuk** [reply]
Thank you!

For some reason, accept.js is not used in my environment. Maybe it's because I use a sandbox account? I will finish the integration on Monday.

Have a nice weekend!

**Reply by delfina-rossi** [reply]
monday is  perfect, thank you so much for the help. Have a nice weekend as well!

**Reply by vovayatsyuk** [reply]
Fixed: https://github.com/breezefront/module-breeze-paradoxlabs-authnetcim/commit/592e458fdeb57a234611b50a06bbcdf942ba9210

**Comment by delfina-rossi** [comment]
Hi Vova!
Wowo! It's working now!! Thank you so much for your assistance.

To summarize, in order for paradox authorize.net module to work,  your integration module is needed (and the patch stated in the repo):
https://github.com/breezefront/module-breeze-paradoxlabs-authnetcim

but also fixes un customer-data.js and region-data.js already added to breeze:

https://github.com/breezefront/module-breeze/commit/1bb89757540f1e0213a74920a28adb87202605e9
https://github.com/breezefront/module-breeze/commit/ee66a4a252d5f7a1bbabc71153f84dfd25a4e755

as well as this one:
https://github.com/breezefront/module-breeze/commit/0e2c8c5887d783848ac80b4bbc55d4be5595b5f8

Thank you for such great support
